WordPress/Classes/ViewRelated/Blog/Blog Dashboard/Service/BlogDashboardPersonalizationService.swift

```diff
@@ -0,0 +1,49 @@
+import Foundation
+
+/// Manages dashboard settings such as card visibility.
+struct BlogDashboardPersonalizationService {
+    private let repository: UserPersistentRepository
+    private let siteID: String
+
+    init(repository: UserPersistentRepository = UserDefaults.standard,
+         siteID: Int) {
+        self.repository = repository
+        self.siteID = String(siteID)
+    }
+
+    func isEnabled(_ card: DashboardCard) -> Bool {
+        getSettings(for: card)[siteID] ?? true
+    }
+
+    func hasPreference(for card: DashboardCard) -> Bool {
+        getSettings(for: card)[siteID] != nil
+    }
+
+    func setEnabled(_ isEnabled: Bool, for card: DashboardCard) {
+        var settings = getSettings(for: card)
+        settings[siteID] = isEnabled
+        repository.set(settings, forKey: makeKey(for: card))
+
+        NotificationCenter.default.post(name: .blogDashboardPersonalizationSettingsChanged, object: self)
+    }
+
+    private func getSettings(for card: DashboardCard) -> [String: Bool] {
+        repository.dictionary(forKey: makeKey(for: card)) as? [String: Bool] ?? [:]
+    }
+}
```

Comment:
I was also debating whether to keep the constants or generate the keys dynamically. The advantage of generating them dynamically is that there are fewer changes needed when new cards that require personalization are added. The disadvantage, as you pointed out, is that they keys are no longer visible in `UPRUConstants`.

> The biggest disadvantage I see is that the raw key is no longer searchable in the codebase.

I can suggest two options how to address it: 
1) Add static keys – one per card – to `UPRUConstants` and create a `DashboardCard` → key mapping
2) Move `makeKey(for card: DashboardCard)` method that I added to `UPRUConstants`, so it's at more visible

I'd be happy to change it based on what you think is the best option. I think making these keys visible/searchable is a good idea.

> The other downside is that "key migration" logic is needed for the prompt key.

The data migration is not needed because I added `.prompts` as an exception:

```swift
private func makeKey(for card: DashboardCard) -> String {
    if card == .prompts {
        // This key was defined statically in the previous versions, and the
        // naming convention wasn't matching the other keys, so it had to be
        // special-cased to avoid losing data.
        return "prompts-enabled-site-settings"
    }
    return "\(card.rawValue)-card-enabled-site-settings"
}
```

I think we can keep it like this indefinitely: it's a small exception and is documented.

Reply:
I think I got your point about searchable. I tried adding the keys to `UPRUConstants`, but there were private, which I didn't want to change. So I added the static strings directly to `makeKey` and retested the migration (or lack thereof) – looks good now. [Commit](https://github.com/wordpress-mobile/WordPress-iOS/pull/20365/commits/2e08a5ec31734e3ea60a786aea748866c0b79638) with the change.